design-documents/checkout/alternative-checkout-flow.md

```diff
@@ -0,0 +1,54 @@
+# Alternative checkout flow
+
+In scope of work on GraphQL and storefront APIs we have an opportunity to improve design and features of storefront checkout.
+
+The purpose of this document is to discuss possible alternatives to current Magento checkout flow. 
+
+  1. When Quote is created?
+     * For physical products on Review & Payments step
+     * For virtual products billing address has to be entered first
+  1. Cart properties:
+     * Line Items:
+       * SKU
+       * Selected options (custom/configurable)
+       * Quantity
+       * Regular price
+       * Price
+  4. Applicable cart rules calculator arguments
+     * Cart
+     * Addresses
+     * Date & Time (admin preview)
+  2. Quote factory arguments: Quote
+     * !Line Items
+     * Dimensions (weight & size) (based on LineItems)
+     * Shipping (requited for physical products)
+       * Address
+       * Selected Shipping Method
+     * Billing address (required for virtual products). Need use cases
+     * Coupons
+     * Gift cards
+     * Store credit
+     * Cart rules (calculated by Applicable catalog rules calculator)
+     * Customer
+  2. Totals calculator: Totals
+     * QuoteArgumentDTO
+     * PreviousTotals
+  5. Quote
+     * LineItems
+       * Totals
+       * Adjustments
+           * Taxes
+           * Cart rule discounts
+     * Billing addresses
+     * Shipping addresses
+     * Coupons
```

Comment:
Why are coupons, gift card codes, store credit are not part of (price) Adjustments? It would make sense for quotes to accept a list of abstract PriceAdjustmentInterface which would be used when calculating totals and to show users which adjustments were applied to their cart/order/quote. I don't think that quote should explicetly rely on taxes, coupons etc. And there must be a mechanism for 3rd-party developers to add their own adjustments. Having an abstract PriceAdjustmentInterface DTO seems that quotes would receive seems like a step to that direction